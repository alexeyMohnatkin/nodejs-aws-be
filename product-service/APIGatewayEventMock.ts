import { APIGatewayProxyEvent } from 'aws-lambda';

const DEFAULT_MOCK: APIGatewayProxyEvent = {
  body: null,
  headers: {},
  multiValueHeaders: {},
  httpMethod: 'GET',
  isBase64Encoded: false,
  path: '',
  pathParameters: null,
  queryStringParameters: null,
  multiValueQueryStringParameters: null,
  stageVariables: null,
  requestContext: null,
  resource: '',
}

export const createAPIGatewayEvent = (event?: Partial<APIGatewayProxyEvent>) => ({
  ...DEFAULT_MOCK,
  ...event,
});
