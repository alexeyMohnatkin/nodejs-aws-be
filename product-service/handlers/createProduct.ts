import { APIGatewayProxyHandler } from 'aws-lambda';
import 'source-map-support/register';

import { Client, ClientConfig } from 'pg';

const { PG_HOST, PG_PORT, PG_DATABASE, PG_USERNAME, PG_PASSWORD } = process.env;

const options: ClientConfig = {
  host: PG_HOST,
  port: +PG_PORT,
  database: PG_DATABASE,
  user: PG_USERNAME,
  password: PG_PASSWORD,
  ssl: {
    rejectUnauthorized: false,
  },
  connectionTimeoutMillis: 5000,
};


const createProduct: APIGatewayProxyHandler = async (event) => {
  console.log('createProduct labmda executed with event: ', event);

  const client = new Client(options);
  await client.connect();

  const createProduct = async () => {
    try {
      await client.query('BEGIN');

      const { title, description, image, price, count } = JSON.parse(event.body);

      const { rows: [{ id: newId }] } = await client.query(`INSERT INTO products (title, description, image, price) VALUES
        ('${title}', '${description}', '${image}', '${price}')
        RETURNING id;
      `);

      await client.query(`INSERT INTO stocks (product_id, count) VALUES
        ('${newId}', ${count});
      `);
      await client.query('COMMIT');
      return newId;

    } catch (error) {
      await client.query('ROLLBACK');
      throw error;
    }
  }

  try {

    const newId = await createProduct();
    const {rows: [product]} = await client.query(`SELECT * FROM products p LEFT JOIN stocks s ON s.product_id = p.id WHERE p.id='${newId}'`);

    return {
      statusCode: 200,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Credentials': true,
      },
      body: JSON.stringify(product),
    };
  } catch (error) {
    console.log(error);
    return {
      statusCode: 500,
      body: JSON.stringify({ message: 'Internal server error' }),
    };
  } finally {
    client.end()
  }
}

export default createProduct;
