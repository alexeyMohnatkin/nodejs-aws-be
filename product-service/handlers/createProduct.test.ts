import { Client } from 'pg';

import { createAPIGatewayEvent } from '../APIGatewayEventMock';

import createProduct from './createProduct';

jest.mock('pg', () => {
  const mClient = {
    connect: jest.fn(),
    query: jest.fn(),
    end: jest.fn(),
  };
  return { Client: jest.fn(() => mClient) };
});

const PRODUCT_PAYLOAD = {
  title: "ProductOne",
  description: "Short Product Description1",
  image: "https://source.unsplash.com/random?sig=0",
  price: 2.4,
  count: 4,
};


describe('Lambda createProduct', () => {
  let client: { connect: jest.Mock, query: jest.Mock, end: jest.Mock };

  beforeAll(() => {
    client = new Client() as any as { connect: jest.Mock, query: jest.Mock, end: jest.Mock };
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should create new product and return it in the response', async () => {
    const newId = '7567ec4b-b10c-48c5-9345-fc73c48a80aa';
    client.query
      .mockResolvedValueOnce(null)
      .mockResolvedValueOnce({rows: [{ id: newId }]})
      .mockResolvedValueOnce(null)
      .mockResolvedValueOnce(null)
      .mockResolvedValueOnce({ rows: [{ ...PRODUCT_PAYLOAD, id: newId }] });

    const event = createAPIGatewayEvent({
      body: JSON.stringify(PRODUCT_PAYLOAD),
    });

    const result = await createProduct(event, null, null);

    const expected = {
      statusCode: 200,
      body: JSON.stringify({ ...PRODUCT_PAYLOAD, id: newId }),
      headers: {
        "Access-Control-Allow-Credentials": true,
        "Access-Control-Allow-Origin": "*",
      },
    };


    expect(client.query.mock.calls).toMatchSnapshot();
    expect(result).toEqual(expected);
  });

  it('should rollback transaction if product creation failed', async () => {
    client.query
      .mockResolvedValueOnce(null)
      .mockRejectedValueOnce('Product creation failed')

    const event = createAPIGatewayEvent({
      body: JSON.stringify(PRODUCT_PAYLOAD),
    });

    const result = await createProduct(event, null, null);

    const expected = {
      statusCode: 500,
      body: JSON.stringify({ message: 'Internal server error' }),
    };

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(result).toEqual(expected);
  });

  it('should rollback transaction if stock creation failed', async () => {
    const newId = '7567ec4b-b10c-48c5-9345-fc73c48a80aa';
    client.query = jest.fn()
      .mockResolvedValueOnce(null)
      .mockResolvedValueOnce({rows: [{ id: newId }]})
      .mockRejectedValueOnce('Stock creation failed')

    const event = createAPIGatewayEvent({
      body: JSON.stringify(PRODUCT_PAYLOAD),
    });

    const result = await createProduct(event, null, null);

    const expected = {
      statusCode: 500,
      body: JSON.stringify({ message: 'Internal server error' }),
    };

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(result).toEqual(expected);
  });
});
