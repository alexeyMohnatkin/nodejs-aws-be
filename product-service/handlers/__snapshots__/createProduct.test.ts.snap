// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda createProduct should create new product and return it in the response 1`] = `
Array [
  Array [
    "BEGIN",
  ],
  Array [
    "INSERT INTO products (title, description, image, price) VALUES
        ('ProductOne', 'Short Product Description1', 'https://source.unsplash.com/random?sig=0', '2.4')
        RETURNING id;
      ",
  ],
  Array [
    "INSERT INTO stocks (product_id, count) VALUES
        ('7567ec4b-b10c-48c5-9345-fc73c48a80aa', 4);
      ",
  ],
  Array [
    "COMMIT",
  ],
  Array [
    "SELECT * FROM products p LEFT JOIN stocks s ON s.product_id = p.id WHERE p.id='7567ec4b-b10c-48c5-9345-fc73c48a80aa'",
  ],
]
`;

exports[`Lambda createProduct should rollback transaction if product creation failed 1`] = `
Array [
  Array [
    "BEGIN",
  ],
  Array [
    "INSERT INTO products (title, description, image, price) VALUES
        ('ProductOne', 'Short Product Description1', 'https://source.unsplash.com/random?sig=0', '2.4')
        RETURNING id;
      ",
  ],
  Array [
    "ROLLBACK",
  ],
]
`;

exports[`Lambda createProduct should rollback transaction if stock creation failed 1`] = `
Array [
  Array [
    "BEGIN",
  ],
  Array [
    "INSERT INTO products (title, description, image, price) VALUES
        ('ProductOne', 'Short Product Description1', 'https://source.unsplash.com/random?sig=0', '2.4')
        RETURNING id;
      ",
  ],
  Array [
    "INSERT INTO stocks (product_id, count) VALUES
        ('7567ec4b-b10c-48c5-9345-fc73c48a80aa', 4);
      ",
  ],
  Array [
    "ROLLBACK",
  ],
]
`;
